### PR TITLE
transaction deserialization no-std env support

### PR DESCRIPTION
**Pull request by zees-dev**
# Context
The core crates don't support `no-std` environments.
This PR adds support for `no-std` to the following crates:
- `xrpl_types`
- `xrpl_binary_codec`

This will allow us to use the respective crates in a `no-std` env (commonly on-chain environments).

---

# Note
- This is required to deserialize encoded transactions on-chain (a `no-std` `wasm32-unknown-unknown` environment).

### CODE REVIEW

**Comment by zees-dev** [comment]
@gmosx Any chance we can merge this?
Or shall I close the PR here.

**Comment by gmosx** [comment]
Hey sorry, I missed this. I will review today and get back to you! Thank you!

**Comment by gmosx** [comment]
@allanbrondumkr can you also check this please?

**Comment by gmosx** [comment]
@zees-dev this test is failing:

test client_tests::tests::client_can_fetch_transactions ... FAILED

can you pls look into it?

**Comment by zees-dev** [comment]
> @zees-dev this test is failing:
> 
> test client_tests::tests::client_can_fetch_transactions ... FAILED
> 
> can you pls look into it?

Will do 👍

**Comment by zees-dev** [comment]
> @zees-dev this test is failing:
> 
> test client_tests::tests::client_can_fetch_transactions ... FAILED
> 
> can you pls look into it?

The test seems to be passing for me:
```sh
cargo test --all-features client_can_fetch_transactions
```

```
running 1 test
test client_tests::tests::client_can_fetch_transactions ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 23 filtered out; finished in 1.00s

     Running unittests src/lib.rs (target/debug/deps/xrpl_types-50397579f17e97f1)
```

---

Edit: The test fails without `--all-features` flag.
Additionally, the test is also failing in `main` branch too... So not introduced by this PR. Maybe previous PR(s)?

**Comment by zees-dev** [comment]
Ok the issue (test breakage) was introduced in [this PR](https://github.com/gmosx/xrpl-sdk-rust/pull/29).
This is because no TLS provider (i think?) is being utilized as we turn off the default features for the `reqwest` library upon compilation.
This PR introduces that specific feature back and defaults the TLS to `tls-default` (which should now allow usage of `https` URLs).
Addressed issue in the [commit](https://github.com/gmosx/xrpl-sdk-rust/pull/31/commits/57950918398a2cbe073134eedd5f107d1a990888) above.

Now `cargo test` should simply work (dont need the `--all-features` flag).

File: xrpl_types/src/types.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
-mod amount;
-mod currency;
-mod currency_code;
-mod ledger_timestamp;
-mod primitive;
-mod transaction;
+pub mod amount;
```
Comment (allanbrondumkr) [inline xrpl_types/src/types.rs]: This means that types have two "paths". Does not seem like a good idea

File: xrpl_types/src/types/transaction/mod.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
-mod common;
-mod variants;
+pub mod common;
```
Comment (allanbrondumkr) [inline xrpl_types/src/types/transaction/mod.rs]: Same here. Seems these changes are unrelated to "no-std" and should be a separate MR (if change is wanted)

Comment (gmosx) [reply xrpl_types/src/types/transaction/mod.rs]: Personally I don't mind about the 'two paths' but to alleviate @allanbrondumkr 's concern @zees-dev could you please remove the `pub` 🙏

Comment (allanbrondumkr) [reply xrpl_types/src/types/transaction/mod.rs]: I may also be wrong, but the crates I know of usually try to have types only at one path if I'm not mistaken. But if there are arguments for a different practice, I would like to understand it.

**Comment by allanbrondumkr** [comment]
> @allanbrondumkr can you also check this please?

I don't know much about what is good practice for supporting "no-std", but changes seems ok. I guess it makes sense to support no-std, although it does add some extra complexity

File: xrpl_types/src/types.rs (hunk 1 above)
Comment (zees-dev) [reply xrpl_types/src/types.rs]: I think the primitives should be exported since they can be used as building blocks.
In fact that was the reason i made them public, since this is a library - it's very probably that the lower level primitives will be composed to create new functionality.
In this case, making them private and inaccessible makes the library difficult to deal with.

File: xrpl_types/src/types/transaction/mod.rs (hunk 2 above)
Comment (zees-dev) [reply xrpl_types/src/types/transaction/mod.rs]: Same comment/reply applies here; in my case i need to use the `TransactionCommon` type (base tx type) from the library.
Unfortunately it's private without this change.

Comment (allanbrondumkr) [reply xrpl_types/src/types/transaction/mod.rs]: The module contents are reexported here:
```
pub use common::*;
pub use variants::*;
```
So making the modules `pub` does not make anything public that was not already? It just makes it available at two paths.

Comment (zees-dev) [reply xrpl_types/src/types/transaction/mod.rs]: o sorry yes i failed to realize that the parent module(s) are already exporting all the child module types.
You are correct, in that scenario we don't need to re-export them (providing multiple paths as you say).
Made a [commit](https://github.com/gmosx/xrpl-sdk-rust/pull/31/commits/9b37448fb1160cd0124ee7a7003ceef0fddf249d) to revert the re-exports.

File: xrpl_types/src/types.rs (hunk 1 above)
Comment (zees-dev) [reply xrpl_types/src/types.rs]: Yep good point, this was not needed. Addressed [here](https://github.com/gmosx/xrpl-sdk-rust/pull/31/commits/9b37448fb1160cd0124ee7a7003ceef0fddf249d) now, Thanks 👍

**Comment by gmosx** [comment]
Thanks for the changes @zees-dev. I will review one more time later today.

**Comment by gmosx** [comment]
Merged, thank you for this contribution @zees-dev
